React/CSS-in-JS/Chakra.md

```diff
@@ -0,0 +1,629 @@
+# Chakra UI
+
+Chakra UI is a simple, modular and accessible component library that gives you the building blocks you need to build your React applications.
+
+## Getting started
+
+### Installation:
+
+```
+npm i @chakra-ui/react @emotion/react @emotion/styled framer-motion
+```
+
+For more about setup read [docs](https://chakra-ui.com/docs/getting-started)
+
+### Setup provider:
+
+For Chakra UI to work correctly, you need to set up the `ChakraProvider` at the root of your application.
+Under the hood Chakra UI is using [emotion ThemeProvider](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/providers.tsx)
+
+Example:
+
+```jsx
+import { ChakraProvider } from "@chakra-ui/react";
+
+function App({ Component, pageProps }: AppProps): ReactElement {
+  return (
+    <ChakraProvider theme={theme}>
+      <Component {...pageProps} />
+    </ChakraProvider>
+  );
+}
+
+export default App;
+```
+
+Note:
+
+- For Next.js, you need to set this up in `pages/_app.tsx`
+- For Create React App, you need to set this up in `index.tsx`
+
+### Style props
+
+Style props are a way to alter the style of a component by simply passing props to it. It helps to save time by providing helpful shorthand ways to style components.
+
+Alternative to that is using sx prop that takes an object with style rules.
```

Comment:
It's fine to mention alternatives, but the preferred way of using it should be clear, together with some examples when the alternative might be better.

Reply:
Style props for one-off styles
```
<Button display="block" mt="md" />
```

`sx` prop when you have to merge multiple objects
```
const ExoticComponent = (props) => {
  const styles = useStyleConfig('ExoticComponent', props);
 
  const defaultStyles = {
   display: 'block',
  };

  return <chakra.div sx={{ ...styles, ...defaultStyles }} />;
}
```
You have control over specificity and merge order. 
Also, there is `__css` which will get merged first, before the `sx`, so I guess we should use `__css` in this case.
https://github.com/chakra-ui/chakra-ui/blob/085891be806b855af90b86367f5b26e8151c3ff5/packages/system/src/system.ts#L106

```
/**
 * Style resolver function that manages how style props are merged
 * in combination with other possible ways of defining styles.
 *
 * For example, take a component defined this way:
 * ```jsx
 * <Box fontSize="24px" sx={{ fontSize: "40px" }}></Box>
 * ```
 *
 * We want to manage the priority of the styles properly to prevent unwanted
 * behaviors. Right now, the `sx` prop has the highest priority so the resolved
 * fontSize will be `40px`
 */
 ```